### PR TITLE
[SPARK-32867][SQL] When explain, HiveTableRelation show limited message

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
In current mode, when explain a SQL plan with HiveTableRelation, it will show so many info about HiveTableRelation's prunedPartition,  this make plan hard to read, this pr make this information simpler.

Before:
image

For UT
```
 test("Make HiveTableScanExec message simple") {
  withSQLConf("hive.exec.dynamic.partition.mode" -> "nonstrict") {
      withTable("df") {
        spark.range(30)
          .select(col("id"), col("id").as("k"))
          .write
          .partitionBy("k")
          .format("hive")
          .mode("overwrite")
          .saveAsTable("df")

        val df = sql("SELECT df.id, df.k FROM df WHERE df.k < 2")
        df.explain(true)
      }
    }
  }
```

After this pr will show
```
== Parsed Logical Plan ==
'Project ['df.id, 'df.k]
+- 'Filter ('df.k < 2)
   +- 'UnresolvedRelation [df], []

== Analyzed Logical Plan ==
id: bigint, k: bigint
Project [id#11L, k#12L]
+- Filter (k#12L < cast(2 as bigint))
   +- SubqueryAlias spark_catalog.default.df
      +- HiveTableRelation [`default`.`df`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, Data Cols: [id#11L], Partition Cols: [k#12L]]

== Optimized Logical Plan ==
Filter (isnotnull(k#12L) AND (k#12L < 2))
+- HiveTableRelation [`default`.`df`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, Data Cols: [id#11L], Partition Cols: [k#12L], Pruned Partitions: [(k=0), (k=1)]]

== Physical Plan ==
Scan hive default.df [id#11L, k#12L], HiveTableRelation [`default`.`df`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, Data Cols: [id#11L], Partition Cols: [k#12L], Pruned Partitions: [(k=0), (k=1)]], [isnotnull(k#12L), (k#12L < 2)]

```

In my pr, I will construct `HiveTableRelation`'s `simpleString` method to avoid show too much unnecessary info in explain plan. compared to what we had before，I decrease the detail metadata of each partition and only retain the partSpec to show each partition was pruned. Since for detail information, we always don't see this in Plan but to use DESC EXTENDED statement.

### Why are the changes needed?
Make plan about HiveTableRelation more readable

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
No

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
cc @wangyum

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by cloud-fan** [comment]
It's only a problem for hive table scan but not the native file source tables?

BTW for hive table only issues, please explain how the problem is resolved in native file source tables, and if it's possible to apply the same approach.

**Comment by AngersZhuuuu** [comment]
For native file source, such as FileSourceScanExec, it will collect metadata of needed information:
```
  override lazy val metadata: Map[String, String] = {
    def seqToString(seq: Seq[Any]) = seq.mkString("[", ", ", "]")
    val location = relation.location
    val locationDesc =
      location.getClass.getSimpleName +
        Utils.buildLocationMetadata(location.rootPaths, maxMetadataValueLength)
    val metadata =
      Map(
        "Format" -> relation.fileFormat.toString,
        "ReadSchema" -> requiredSchema.catalogString,
        "Batched" -> supportsColumnar.toString,
        "PartitionFilters" -> seqToString(partitionFilters),
        "PushedFilters" -> seqToString(pushedDownFilters),
        "DataFilters" -> seqToString(dataFilters),
        "Location" -> locationDesc)

    val withSelectedBucketsCount = relation.bucketSpec.map { spec =>
      val numSelectedBuckets = optionalBucketSet.map { b =>
        b.cardinality()
      } getOrElse {
        spec.numBuckets
      }
      metadata + ("SelectedBucketsCount" ->
        (s"$numSelectedBuckets out of ${spec.numBuckets}" +
          optionalNumCoalescedBuckets.map { b => s" (Coalesced to $b)"}.getOrElse("")))
    } getOrElse {
      metadata
    }

    withSelectedBucketsCount
  }
```
Then in his parent class  `DataSourceScanExec`, it will construct `simpleString` with `metadata`. In `simpleString`
```
  override def simpleString(maxFields: Int): String = {
    val metadataEntries = metadata.toSeq.sorted.map {
      case (key, value) =>
        key + ": " + StringUtils.abbreviate(redact(value), maxMetadataValueLength)
    }
    val metadataStr = truncatedString(metadataEntries, " ", ", ", "", maxFields)
    redact(
      s"$nodeNamePrefix$nodeName${truncatedString(output, "[", ",", "]", maxFields)}$metadataStr")
  }
```
each info will be abbreviated to length of 100. So FileSourceScan's  FileIndex message will be hided because of length of 100.

**Comment by AngersZhuuuu** [comment]
@cloud-fan As mentioned above, I use similar way to construct HiveTableRelation's simpleString.  Compared to what we had before，I decrease the detail metadata of each partition and only retain the partSpec to show each partition was pruned.
since for detail information, we always don't see this in Plan but to use `DESC EXTENDED` statement.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 1]
```diff
@@ -55,7 +55,7 @@ trait DataSourceScanExec extends LeafExecNode {
   // Metadata that describes more details of this scan.
   protected def metadata: Map[String, String]
 
-  protected val maxMetadataValueLength = 100
+  protected val maxMetadataValueLength = conf.maxMetadataValueLength
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:58]: Have you checked the related PR? https://github.com/apache/spark/pull/29688/files#diff-2a91a9a59953aa82fa132aaf45bd731bR58

**Comment by maropu** [comment]
> As mentioned above, I use similar way to construct HiveTableRelation's simpleString.

Please update the PR description about how-to-fix.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:58]: > Have you checked the related PR? https://github.com/apache/spark/pull/29688/files#diff-2a91a9a59953aa82fa132aaf45bd731bR58

Sorry, don't notice that pr. I have similar ideal like that pr too.  I'm going to restore this change and deal with the two PR conflicts  finally.

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:58]: link: https://github.com/apache/spark/pull/29688

**Comment by maropu** [comment]
Could you add tests, too?

**Comment by AngersZhuuuu** [comment]
> Could you add tests, too?

Sure, just check HiveTableScanExec's simpleString will be ok?

**Comment by AngersZhuuuu** [comment]
> Please update the PR description about how-to-fix.

Done

**Comment by AngersZhuuuu** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 2]
```diff
@@ -693,4 +694,29 @@ case class HiveTableRelation(
   override def newInstance(): HiveTableRelation = copy(
     dataCols = dataCols.map(_.newInstance()),
     partitionCols = partitionCols.map(_.newInstance()))
+
+  override def simpleString(maxFields: Int): String = {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:698]: The explain output of `HiveTableScanExec` does not have the same issue?

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:698]: > The explain output of `HiveTableScanExec` does not have the same issue?

No, only caused. by `HiveTableRelation`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 3]
```diff
@@ -693,4 +694,29 @@ case class HiveTableRelation(
...
+    val catalogTable = tableMeta.storage.serde match {
+      case Some(serde) => tableMeta.identifier :: serde :: Nil
+      case _ => tableMeta.identifier :: Nil
+    }
+
+    val metadataEntries = Map(
+      "CatalogTable" -> catalogTable.mkString(", "),
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:705]: is it also how file source tables display table info?

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:705]: > is it also how file source tables display table info?

File source tables won't show this.
```
== Parsed Logical Plan ==
'Project [*]
+- 'Filter ('df.k < 3)
   +- 'UnresolvedRelation [df], []

== Analyzed Logical Plan ==
id: bigint, k: bigint
Project [id#7L, k#8L]
+- Filter (k#8L < cast(3 as bigint))
   +- SubqueryAlias spark_catalog.default.df
      +- Relation[id#7L,k#8L] parquet

== Optimized Logical Plan ==
Filter (isnotnull(k#8L) AND (k#8L < 3))
+- Relation[id#7L,k#8L] parquet

== Physical Plan ==
*(1) ColumnarToRow
+- FileScan parquet default.df[id#7L,k#8L] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[file:/Users/angerszhu/Documents/project/AngersZhu/spark/sql/hive/target/tmp/hiv..., PartitionFilters: [isnotnull(k#8L), (k#8L < 3)], PushedFilters: [], ReadSchema: struct<id:bigint>

```

**Comment by AngersZhuuuu** [comment]
retest this please

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala [hunk 4]
```diff
@@ -187,6 +189,47 @@ class HiveTableScanSuite extends HiveComparisonTest with SQLTestUtils with TestH
...
+          .partitionBy("k")
+          .format("hive")
+          .mode("overwrite")
+          .saveAsTable("df")
+
+        val scan1 = getHiveTableScanExec("SELECT * FROM df WHERE df.k < 3")
+        assert(scan1.simpleString(100).replaceAll("#\\d+L", "") ==
+          "Scan hive default.df [id, k]," +
+            " HiveTableRelation [" +
+            "`default`.`df`," +
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala:208]: This seems duplicated to `Scan hive default.df`

Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala:208]: > This seems duplicated to `Scan hive default.df`

Yea, in origin plan, also have this problem. we can't drop this in `HiveTableRelation`  since we need to show table in Optimized plan and  analyzed plan. 

How about  implement HiveTableScanExec's `simpleString` to filter  HiveTableRelation's table info?

Comment (cloud-fan) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala:208]: let's leave it, as it's a trivial problem.

Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala:208]: > let's leave it, as it's a trivial problem.

Ok.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 5]
```diff
@@ -693,4 +694,29 @@ case class HiveTableRelation(
...
+      "Data Cols" -> truncatedString(dataCols, "[", ", ", "]", maxFields),
+      "Partition Cols" -> truncatedString(partitionCols, "[", ", ", "]", maxFields),
+      "Statistic" -> tableStats.map(_.simpleString).getOrElse(""),
+      "Pruned Partitions" -> prunedPartitions.map { partitions =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:709]: AFAIK, in Hive each partition can have a different serde. Shall we display partition serde like before?

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:709]: > AFAIK, in Hive each partition can have a different serde. Shall we display partition serde like before?

Since table level `serde` wan shown, show partition's `serde` is ok for me.

But be honest, default meta value length is 100, one partition `serde` length always more than 30. It won't show too much info....

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:709]: maybe omit the partition serde if it's the same as table serde?

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:709]: > maybe omit the partition serde if it's the same as table serde?

Looks good to me. update later

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:709]: > maybe omit the partition serde if it's the same as table serde?

Strictly speaking, the field information for the partition may also be different

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala [hunk 6]
```diff
@@ -187,6 +189,47 @@ class HiveTableScanSuite extends HiveComparisonTest with SQLTestUtils with TestH
     }
   }
 
+  test("SPARK-32867: When explain, HiveTableRelation show limited message") {
+    withSQLConf(HiveUtils.CONVERT_METASTORE_ORC.key -> "false",
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala:193]: Is this config related to this test? Looks we can write tests without it though.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala [hunk 7]
```diff
@@ -187,6 +189,47 @@ class HiveTableScanSuite extends HiveComparisonTest with SQLTestUtils with TestH
...
+        val scan1 = getHiveTableScanExec("SELECT * FROM df WHERE df.k < 3")
+        assert(scan1.simpleString(100).replaceAll("#\\d+L", "") ==
+          "Scan hive default.df [id, k]," +
+            " HiveTableRelation [" +
+            "`default`.`df`," +
+            " org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe," +
+            " Data Cols: [id]," +
+            " Partition Cols: [k]," +
+            " Pruned Partitions: [k=0, k=1, k=2]," +
+            " Statistic: sizeInBytes=8.0 EiB" +
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala:213]: Printing `8.0 EiB` looks weird and we need `Statistic` here?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala [hunk 8]
```diff
@@ -187,6 +189,47 @@ class HiveTableScanSuite extends HiveComparisonTest with SQLTestUtils with TestH
     }
   }
 
+  test("SPARK-32867: When explain, HiveTableRelation show limited message") {
+    withSQLConf(HiveUtils.CONVERT_METASTORE_ORC.key -> "false",
+      "hive.exec.dynamic.partition.mode" -> "nonstrict") {
+      withTable("df") {
+        spark.range(1000)
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala:196]: > I think we don't need to make many partitions, so `1000` -> `4`?

how about 30? since I want to show `....` for HiveTableRelation..

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala (hunk 7 above)
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala:213]: > Printing `8.0 EiB` looks weird and we need `Statistic` here?

Not so important, IMO, we can remove

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (maropu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala:196]: sgtm

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala (hunk 6 above)
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala]: > Is this config related to this test? Looks we can write tests without it though.

Just a  habit about writing UT of hive table. Remove it, thanks for correct this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala (hunk 5 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: > maybe omit the partition serde if it's the same as table serde?

Updated

**Comment by AngersZhuuuu** [comment]
Any update? ping @cloud-fan @maropu

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by maropu** [comment]
Thanks for this fix, @AngersZhuuuu
